### PR TITLE
Persist config

### PR DESCRIPTION
**Pull request by joaquimrocha**
* Update runtimes/shared-modules;
* Make preferences persistent.

### CODE REVIEW

**Comment by joaquimrocha** [comment]
bot, build com.abagames.noiz2sa

**Comment by bbhtt** [comment]
Hello, are you working on this? Would be nice to have it on the latest runtime. It seems you need a 128x128 icon.

File: noiz2sa.desktop [hunk 1]
```diff
@@ -3,6 +3,6 @@ Type=Application
 Name=Noiz2sa
 Comment=An Abstract Shooter by Kenta Cho
 Icon=noiz2sa
-Exec=/app/bin/noiz2sa
+Exec=/app/bin/run-noiz2sa.sh
```
Comment (hfiguiere) [inline noiz2sa.desktop]:
```suggestion
Exec=run-noiz2sa.sh
```

File: com.abagames.noiz2sa.yaml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    commands:
+    - install -Dm644 noiz2sa.appdata.xml /app/share/appdata/noiz2sa.appdata.xml
+  - type: script
+    dest-filename: run-noiz2sa.sh
+    commands:
+    - !/bin/sh -e
+    - HOME=$XDG_CONFIG_HOME /app/bin/noiz2sa
+  build-commands:
+  - install ./noiz2sa /app/bin/noiz2sa
+  - install -m 555 ../run-noiz2sa.sh /app/bin/run-noiz2sa.sh
```
Comment (hfiguiere) [inline com.abagames.noiz2sa.yaml:87]:
```suggestion
    - HOME=$XDG_CONFIG_HOME /app/bin/noiz2sa.bin
  build-commands:
  - install ./noiz2sa /app/bin/noiz2sa.bin
  - install -m 555 ../run-noiz2sa.sh /app/bin/run-noiz2sa
```

That way you don't have to change the executable in the .desktop, or the `command` in the manifest.

File: noiz2sa.desktop (hunk 1 above)
Comment (hfiguiere) [reply noiz2sa.desktop]: just to clarify: `Exec` isn't a full-path, regardless of the other suggestion.

File: com.abagames.noiz2sa.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+command: run-noiz2sa.sh
+rename-desktop-file: noiz2sa.desktop
+rename-icon: noiz2sa
+rename-appdata-file: noiz2sa.appdata.xml
+finish-args:
+- "--device=dri"
+- "--share=ipc"
+- "--socket=pulseaudio"
+- "--socket=x11"
+- "--persist=.noiz2sa.prf"
```
Comment (hfiguiere) [inline com.abagames.noiz2sa.yaml]: this is likely no longer needed.

File: com.abagames.noiz2sa.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
+app-id: com.abagames.noiz2sa
+runtime: org.freedesktop.Platform
+runtime-version: '20.08'
```
Comment (hfiguiere) [inline com.abagames.noiz2sa.yaml]: you might as well update the runtime to `21.08`. 20.08 will be EOL in August.

File: com.abagames.noiz2sa.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    path: patches/bulletml/get-rid-of-boost.patch
+  build-commands:
+  - make -C src CXXFLAGS="$(CXXFLAGS) $(CPPFLAGS) -fPIC -fpic" LDFLAGS="$(LDFLAGS)
+    -Wl,-z,defs" libbulletml.so
+  - mkdir -p /app/include/d/ /app/include/bulletml/ /app/include/bulletml/tinyxml/
+    /app/lib
+  - cp src/bulletml.d /app/include/d/
+  - cp src/*.h /app/include/bulletml/
+  - cp src/tinyxml/*.h /app/include/bulletml/tinyxml/
+  - cp src/lib*.so* /app/lib/
```
Comment (hfiguiere) [inline com.abagames.noiz2sa.yaml]: there should be a `cleanup` rule to remove the content of `/include` after the build.

**Comment by bbhtt** [comment]
@joaquimrocha Hi, are you still working on this?

**Comment by joaquimrocha** [comment]
Hey. I think for some reason I didn't see all these updates.
I have a bit of time this weekend, so I'll try to take a look.

**Comment by joaquimrocha** [comment]
I'm also happy to add others as owners of this project.

File: com.abagames.noiz2sa.yaml (hunk 4 above)
Comment (bbhtt) [reply com.abagames.noiz2sa.yaml]: It seems to run with 23.08 https://github.com/flathub/com.abagames.noiz2sa/pull/2, please try 23.08. Feel free to take the changes from that PR to here.

**Comment by joaquimrocha** [comment]
I will solve the conflict later.

**Comment by joaquimrocha** [comment]
@bbhtt , I have finally updated it. If you don't mind, please review again.

**Comment by hfiguiere** [comment]
> I have finally updated it. If you don't mind, please review again.

If you want **volunteer** to review, and you decided to convert from one format to another, you are basically being rude to them. Because it make this **unreviewable**.

File: com.abagames.noiz2sa.yaml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  - type: file
+    path: noiz2sa.desktop
+  - type: shell
+    commands:
+    - install -Dm644 noiz2sa.desktop /app/share/applications/noiz2sa.desktop
+  - type: file
+    path: noiz2sa.png
+  - type: shell
+    commands:
+    - install -Dm644 noiz2sa.png /app/share/icons/hicolor/64x64/apps/noiz2sa.png
```
Comment (hfiguiere) [inline com.abagames.noiz2sa.yaml]: minimum size required is 128x128. That's probably while the build failed.

Comment (bbhtt) [reply com.abagames.noiz2sa.yaml]: I had fixed this in https://github.com/flathub/com.abagames.noiz2sa/commit/1324585b5cf7ce6a48f7846dcca65d0fe26e32ff, probably lost in the yaml conversion

Comment (joaquimrocha) [reply com.abagames.noiz2sa.yaml]: I didn't notice that. I thought I had ported all the changed from your PR. Sorry.

**Comment by joaquimrocha** [comment]
> > I have finally updated it. If you don't mind, please review again.
> 
> If you want **volunteer** to review, and you decided to convert from one format to another, you are basically being rude to them. Because it make this **unreviewable**.

I don't understand what you mean. This PR had already the change to YAML if that's what you mean.
Judging from your other comment, I probably forgot to port some of the current master changes after the other PR merge yesterday, but I didn't mean to be rude or make things more difficult to review. In fact, I thought that by merging the other PR before, it'd make things easier to be reviewed.

**Comment by bbhtt** [comment]
> I don't understand what you mean. This PR had already the change to YAML if that's what you mean.

YAML-ing makes the diff harder to (re)view as it is a new manifest entirely instead of changes on top of the existing one. Is there any reason you're yamling it? There's not really much benefit over json.

Fwiw I didn't take this as rude :)

**Comment by joaquimrocha** [comment]
My intention was to have a plain YAML conversion in that first commit. If I failed to port something, that was a mistake and I still need to look closer into it (and the Robocode changes too).

About why YAML: I just prefer it to JSON as I think it's easier to read and change. I didn't think this could be a controversial change 😊

**Comment by joaquimrocha** [comment]
I have re-run the conversion of JSON to YAML from master's JSON instead of trying to add the changes manually as before.
This means the conversion commit doesn't need much attention and nothing except the format has changed.
The config persistency is thus the only functional changes in this PR. Hope it's okay now and I apologize again for not having properly updated things last time.

File: com.abagames.noiz2sa.yaml (hunk 5 above)
Comment (bbhtt) [reply com.abagames.noiz2sa.yaml]: @joaquimrocha Move the cleanup to a global one. SDL leaves some leftovers in `/include`. Also add `/lib/pkgconfig`, `/lib/*.la`

I'll fix this in the shared-modules later.

**Comment by bbhtt** [comment]
Script to convert:

```python
import ruamel.yaml
import json

in_file = 'com.abagames.noiz2sa.yaml'
out_file = 'output.json'

yaml = ruamel.yaml.YAML(typ='safe')
with open(in_file) as file_in:
    data = yaml.load(file_in)
with open(out_file, 'w') as file_out:
    json.dump(data, file_out, indent=2)
```
Diff between original manifest and converted:

```diff
--- com.abagames.noiz2sa.json	2023-09-10 07:18:27.475290460 +0530
+++ output.json	2023-09-10 07:29:22.706708863 +0530
@@ -3,7 +3,7 @@
   "runtime": "org.freedesktop.Platform",
   "runtime-version": "23.08",
   "sdk": "org.freedesktop.Sdk",
-  "command": "noiz2sa",
+  "command": "run-noiz2sa.sh",
   "rename-desktop-file": "noiz2sa.desktop",
   "rename-icon": "noiz2sa",
   "rename-appdata-file": "noiz2sa.appdata.xml",
@@ -11,8 +11,7 @@
     "--device=dri",
     "--share=ipc",
     "--socket=pulseaudio",
-    "--socket=x11",
-    "--persist=.noiz2sa.prf"
+    "--socket=x11"
   ],
   "modules": [
     "shared-modules/glu/glu-9.json",
@@ -63,6 +62,9 @@
         "cp src/*.h /app/include/bulletml/",
         "cp src/tinyxml/*.h /app/include/bulletml/tinyxml/",
         "cp src/lib*.so* /app/lib/"
+      ],
+      "cleanup": [
+        "/include"
       ]
     },
     {
@@ -70,7 +72,10 @@
       "no-autogen": true,
       "no-make-install": true,
       "subdir": "src",
-      "make-args": ["-f", "makefile.lin"],
+      "make-args": [
+        "-f",
+        "makefile.lin"
+      ],
       "sources": [
         {
           "type": "archive",
@@ -118,13 +123,21 @@
           "commands": [
             "install -Dm644 noiz2sa.appdata.xml /app/share/appdata/noiz2sa.appdata.xml"
           ]
+        },
+        {
+          "type": "script",
+          "dest-filename": "run-noiz2sa.sh",
+          "commands": [
+            "HOME=$XDG_CONFIG_HOME /app/bin/noiz2sa"
+          ]
         }
       ],
       "build-commands": [
         "install ./noiz2sa /app/bin/noiz2sa",
+        "install -m 555 ../run-noiz2sa.sh /app/bin/run-noiz2sa.sh",
         "install -d /app/share/games/noiz2sa",
         "cp -R ../noiz2sa_share/* /app/share/games/noiz2sa"
       ]
     }
   ]
-}
+}
\ No newline at end of file
```

File: com.abagames.noiz2sa.yaml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+- "--socket=x11"
+modules:
+- shared-modules/glu/glu-9.json
+- shared-modules/SDL/SDL-1.2.15.json
+- shared-modules/SDL/SDL_mixer-1.2.12.json
+- name: bulletml
+  buildsystem: simple
+  sources:
+  - type: archive
+    url: http://http.debian.net/debian/pool/main/b/bulletml/bulletml_0.0.6.orig.tar.gz
```
Comment (bbhtt) [inline com.abagames.noiz2sa.yaml:22]: `ldconfig: /app/lib/libbulletml.so.0d2 is not a symbolic link` while running.

Looks like this symlink is not created: `libbulletml.so.0d2 -> libbulletml.so.0d2.0`

The patches looks like they are from debian which has the symlink. Can you please check if the source or any of them needs updating to fix this? Thanks.

File: com.abagames.noiz2sa.yaml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    - install -Dm644 noiz2sa.png /app/share/icons/hicolor/128x128/apps/noiz2sa.png
+  - type: file
+    path: noiz2sa.appdata.xml
+  - type: shell
+    commands:
+    - install -Dm644 noiz2sa.appdata.xml /app/share/appdata/noiz2sa.appdata.xml
+  - type: script
+    dest-filename: run-noiz2sa.sh
+    commands:
+    - !/bin/sh -e
```
Comment (bbhtt) [inline com.abagames.noiz2sa.yaml]: The `type: script` sets up a shebang.

```
[📦 com.abagames.noiz2sa ~]$ cat /app/bin/run-noiz2sa.sh 
#!/bin/sh
HOME=$XDG_CONFIG_HOME /app/bin/noiz2sa
```

Yours is:

```
[📦 com.abagames.noiz2sa ~]$ cat /app/bin/run-noiz2sa.sh  
#!/bin/sh
-e
HOME=$XDG_CONFIG_HOME /app/bin/noiz2sa
```

File: com.abagames.noiz2sa.yaml (hunk 5 above)
Comment (bbhtt) [reply com.abagames.noiz2sa.yaml]: > I'll fix this in the shared-modules later.

Merged now, you just need to update the shared-modules submodule. Keep rest as is.

File: com.abagames.noiz2sa.yaml (hunk 7 above)
Comment (bbhtt) [reply com.abagames.noiz2sa.yaml:22]: Ah found the issue, you're doing plain `cp` in the build commands which doesn't preserve the link, change it to `cp -a` `cp -a src/lib*.so* /app/lib/`

**Comment by joaquimrocha** [comment]
Thanks for the review @bbhtt and for investigating the causes of those issues.

About this one though:
> Script to convert:
> ...
> Diff between original manifest and converted:

This is because you've checked the YAML after the 2nd commit. The reason why I have two commits is precisely to make the conversion independent from the persistent config.
Running your script just after the 1st commit of this PR yields a diff like:
```diff
git diff master:com.abagames.noiz2sa.json output.json
diff --git a/com.abagames.noiz2sa.json b/output.json
index ed32e36..07b7b73 100644
--- a/com.abagames.noiz2sa.json
+++ b/output.json
@@ -70,7 +70,10 @@
       "no-autogen": true,
       "no-make-install": true,
       "subdir": "src",
-      "make-args": ["-f", "makefile.lin"],
+      "make-args": [
+        "-f",
+        "makefile.lin"
+      ],
       "sources": [
         {
           "type": "archive",
@@ -127,4 +130,4 @@
       ]
     }
   ]
-}
+}
\ No newline at end of file
```

i.e. there are no functional changes in the conversion.

I will update the 2nd commit to add the fixes you've suggested. Thank you again.

**Comment by joaquimrocha** [comment]
I have updated the branch now and added an extra commit for the "preserve all" option for cp. I appreciate your help @bbhtt .

**Review by bbhtt — APPROVED** [review]

**Pull request merged**